core/trino-main/src/main/java/io/trino/sql/planner/rowpattern/IrRowPatternFlattener.java

```diff
@@ -0,0 +1,242 @@
+                        if (child instanceof IrAlternation) {
+                            return ((IrAlternation) child).getPatterns().stream();
+                        }
+                        else {
+                            return Stream.of(child);
+                        }
+                    })
+                    .collect(toImmutableList());
+
+            Optional<IrRowPattern> firstEmptyChild = children.stream()
+                    .filter(IrEmpty.class::isInstance)
+                    .findFirst();
+            if (firstEmptyChild.isEmpty()) {
+                return new IrAlternation(children);
+            }
+
+            // remove all empty children following the first empty child
+            children = children.stream()
+                    .filter(child -> !(child instanceof IrEmpty) || child == firstEmptyChild.get())
+                    .collect(toImmutableList());
+
+            // if there is only the empty child left, replace alternation with empty pattern
+            if (children.size() == 1) {
+                return new IrEmpty();
+            }
+
+            return new IrAlternation(children);
+        }
+
+        /**
+         * Flatten the concatenation and remove all empty branches.
+         * This method recursively inlines sub-pattern lists from child concatenations into the top-level concatenation.
+         * The expected sub-pattern order is reflected in the resulting order of the elements.
+         * Also, all empty branches are removed from the resulting sub-pattern list.
+         * Examples:
+         * A (B C) -> A B C
+         * (A B) ((C D) E) -> A B C D E
+         * A (() B) -> A B
+         * (A ()) ((() B) ()) -> A B
+         * () (() A) -> A
+         * (() ()) () -> ()
+         *
+         * @return the flattened IrConcatenation containing no empty branches, or a sub-pattern in case when
+         * the concatenation is reduced to a single branch, or IrEmpty in case when all sub-patterns
+         * are empty
+         */
+        @Override
+        protected IrRowPattern visitIrConcatenation(IrConcatenation node, Boolean inExclusion)
+        {
+            List<IrRowPattern> children = node.getPatterns().stream()
+                    .map(pattern -> process(pattern, inExclusion))
+                    .collect(toImmutableList());
+
+            // flatten concatenation
+            ImmutableList.Builder<IrRowPattern> builder = ImmutableList.builder();
+            children.stream()
+                    .forEach(child -> {
+                        if (child instanceof IrConcatenation) {
+                            builder.addAll(((IrConcatenation) child).getPatterns());
+                        }
+                        else {
+                            builder.add(child);
+                        }
+                    });
+            children = builder.build();
+
+            // remove all empty children
+            children = children.stream()
```

Comment:
After you change the above to use flatMap, this chain can be combined into the previous stream() chain.